src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/RazorDocumentOnTypeFormattingEndpoint.cs

```diff
@@ -0,0 +1,180 @@
+
+            return documentSnapshot;
+        }, cancellationToken).ConfigureAwait(false);
+
+        if (documentSnapshot is null)
+        {
+            _logger.LogWarning("Failed to find document {requestTextDocumentUri}.", request.TextDocument.Uri);
+            return null;
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+        if (codeDocument.IsUnsupported())
+        {
+            _logger.LogWarning("Failed to retrieve generated output for document {requestTextDocumentUri}.", request.TextDocument.Uri);
+            return null;
+        }
+
+        var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+        if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var hostDocumentIndex))
+        {
+            return null;
+        }
+
+        var triggerCharacterKind = _razorDocumentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex, rightAssociative: false);
+        if (triggerCharacterKind is not (RazorLanguageKind.CSharp or RazorLanguageKind.Html))
+        {
+            _logger.LogInformation("Unsupported trigger character language {triggerCharacterKind:G}.", triggerCharacterKind);
+            return null;
+        }
+
+        if (!IsApplicableTriggerCharacter(request.Character, triggerCharacterKind))
+        {
+            // We were triggered but the trigger character doesn't make sense for the current cursor position. Bail.
+            _logger.LogInformation("Unsupported trigger character location.");
+            return null;
+        }
+
+        cancellationToken.ThrowIfCancellationRequested();
+
+        Debug.Assert(request.Character.Length > 0);
+
+        var formattedEdits = await _razorFormattingService.FormatOnTypeAsync(request.TextDocument.Uri, documentSnapshot, triggerCharacterKind, Array.Empty<TextEdit>(), request.Options, hostDocumentIndex, request.Character[0], cancellationToken).ConfigureAwait(false);
+        if (formattedEdits.Length == 0)
+        {
+            _logger.LogInformation("No formatting changes were necessary");
+            return null;
+        }
+
+        _logger.LogInformation("Returning {formattingEditsLength} final formatted results.", formattedEdits.Length);
+        return formattedEdits;
+    }
+
+    private static bool IsApplicableTriggerCharacter(string triggerCharacter, RazorLanguageKind languageKind)
+    {
+        if (languageKind == RazorLanguageKind.CSharp)
+        {
+            return s_csharpTriggerCharacters.Contains(triggerCharacter);
+        }
+        else if (languageKind == RazorLanguageKind.Html)
+        {
+            return s_htmlTriggerCharacters.Contains(triggerCharacter);
+        }
+
+        // Unknown trigger character.
+        return false;
+    }
+
+    public RegistrationExtensionResult? GetRegistration(VSInternalClientCapabilities clientCapabilities)
```

Comment:
For all the formatting endpoints could we put the `GetRegistration` methods above the `Handle` calls? Makes understanding capabilities a lot easier / simpler at a glance

Reply:
Totally true that they're less likely to be modified. From a readability perspective I typically put them up top though because it makes it super easy to see what we do and don't support which at least for me has been the usual question folks have asked in the past